lib/elixir/lib/calendar/duration.ex

```diff
@@ -0,0 +1,173 @@
+defmodule Duration do
+  @moduledoc """
+  Struct and functions for handling durations.
+
+  A `Duration` struct represents a collection of time scale units,
+  allowing for manipulation and calculation of durations.
+
+  Date and time scale units are represented as integers, allowing for both positive and negative values.
+
+  Microseconds are represented using a tuple `{microsecond, precision}`.
+  This ensures compatibility with other calendar types implementing time, such as `Time`, `DateTime`, and `NaiveDateTime`.
+  """
+
+  defstruct year: 0,
+            month: 0,
+            week: 0,
+            day: 0,
+            hour: 0,
+            minute: 0,
+            second: 0,
+            microsecond: {0, 0}
+
+  @type t :: %Duration{
+          year: integer,
+          month: integer,
+          week: integer,
+          day: integer,
+          hour: integer,
+          minute: integer,
+          second: integer,
+          microsecond: {integer, integer}
+        }
+
+  @type unit ::
+          {:year, integer}
+          | {:month, integer}
+          | {:week, integer}
+          | {:day, integer}
+          | {:hour, integer}
+          | {:minute, integer}
+          | {:second, integer}
+          | {:microsecond, {integer, integer}}
+
+  @doc """
+  Creates a new `Duration` struct from given `units`.
+
+  Raises a KeyError when called with invalid units.
+
+  ## Examples
+
+      iex> Duration.new(month: 2)
+      %Duration{month: 2}
+
+  """
+  @spec new([unit]) :: t
+  def new(units) do
+    case Keyword.get(units, :microsecond) do
+      nil ->
+        :noop
+
+      ms when is_tuple(ms) ->
+        :noop
+
+      _ ->
+        raise "microseconds must be a tuple {ms, precision}"
```

Comment:
Do we need to change the function `new` to `new!` because it raises the exception.

Reply:
Given it raises on user input only, i think it's fine to be a non-bang version. Similar to `Map.new(123)` raising as well

Reply:
Just for transparency, this came up again [here](https://github.com/elixir-lang/elixir/pull/13385#discussion_r1550659497) and we've decided to be consistent with `new` on other calendar types regardless of what i wrote above :)